core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala

```diff
@@ -0,0 +1,939 @@
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package kafka.coordinator
+
+import java.util.concurrent.locks.ReentrantReadWriteLock
+
+import kafka.utils.CoreUtils._
+import org.apache.kafka.common.protocol.Errors
+import org.apache.kafka.common.protocol.types.{ArrayOf, Struct, Schema, Field}
+import org.apache.kafka.common.protocol.types.Type.STRING
+import org.apache.kafka.common.protocol.types.Type.INT32
+import org.apache.kafka.common.protocol.types.Type.INT64
+import org.apache.kafka.common.protocol.types.Type.BYTES
+import org.apache.kafka.common.utils.Utils
+
+import kafka.utils._
+import kafka.common._
+import kafka.message._
+import kafka.log.FileMessageSet
+import kafka.metrics.KafkaMetricsGroup
+import kafka.common.TopicAndPartition
+import kafka.tools.MessageFormatter
+import kafka.api.ProducerResponseStatus
+import kafka.server.ReplicaManager
+
+import scala.collection._
+import java.io.PrintStream
+import java.nio.ByteBuffer
+import java.util.concurrent.atomic.AtomicBoolean
+import java.util.concurrent.{CountDownLatch, TimeUnit}
+
+import com.yammer.metrics.core.Gauge
+
+class GroupMetadataManager(val brokerId: Int,
+                           val config: OffsetConfig,
+                           replicaManager: ReplicaManager,
+                           zkUtils: ZkUtils,
+                           scheduler: Scheduler) extends Logging with KafkaMetricsGroup {
+
+  /* offsets cache */
+  private val offsetsCache = new Pool[GroupTopicPartition, OffsetAndMetadata]
+
+  /* group metadata cache */
+  private val groupsCache = new Pool[String, GroupMetadata]
+
+  /* partitions of consumer groups that are being loaded, its lock should be always called BEFORE the group lock if needed */
+  private val loadingPartitions: mutable.Set[Int] = mutable.Set()
+
+  /* partitions of consumer groups that are assigned, using the same loading partition lock */
+  private val ownedPartitions: mutable.Set[Int] = mutable.Set()
+
+  /* lock for expiring stale offsets, it should be always called BEFORE the group lock if needed */
+  private val offsetExpireLock = new ReentrantReadWriteLock()
+
+  /* lock for removing offsets of a range partition, it should be always called BEFORE the group lock if needed */
+  private val offsetRemoveLock = new ReentrantReadWriteLock()
```

Comment:
This lock is used to protect race condition when group migration happens at the same time during get offsets: if the group is removed after the ```isGroupLocal``` check, the coordinator with return empty offsets with NONE error code, instead of the right behavior of returning NOT_COORDINATOR error code, this will result in client side handling it as "we do not have any offsets stored in Kafka".

Actually I think this does expose a bug in the code: we need to include the removal of offsets also in the loadingPartition object lock, and hence we can remove this lock. @junrao Thoughts?